dbms/src/Storages/Page/V3/PageDirectory.cpp

```diff
@@ -1504,6 +1505,12 @@ void PageDirectory<Trait>::apply(PageEntriesEdit && edit, const WriteLimiterPtr
                 case EditRecordType::UPDATE_DATA_FROM_REMOTE:
                     throw Exception(fmt::format("should not handle edit with invalid type [type={}]", magic_enum::enum_name(r.type)));
                 }
+
+                // collect the applied remote data_file_ids
+                if (r.entry.checkpoint_info)
```

Comment:
Is this change used in restoring from WAL / Snapshot?

Reply:
it is used for removing the duplicated ids in `S3LockLocalManager::pre_lock_keys`

Before a write batch is applied to PageDirectory, `CPFilesWriter::writeEditsAndApplyCheckpointInfo` can not know the existence of these locks, so we store them into `S3LockLocalManager::pre_lock_keys`. When a checkpoint is going to upload, we set them to `DumpCheckpointOptions::must_locked_files`.
After the write batch is applied, we can remove these locks from `S3LockLocalManager::pre_lock_keys`.

Otherwise, we can not know when to remove the keys in `S3LockLocalManager::pre_lock_keys`.